patroni/validator.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+import logging
+import os
+import socket
+
+from patroni.utils import split_host_port
+
+logger = logging.getLogger(__name__)
+
+
+def data_directory_empty(data_dir):
+    if os.path.isfile(os.path.join(data_dir, "global", "pg_control")):
+        return False
+    if not os.path.exists(data_dir):
+        return True
+    return all(
+        os.name != "nt" and (n.startswith(".") or n == "lost+found")
+        for n in os.listdir(data_dir)
+    )
+
+def which(program):
+    for path in os.environ["PATH"].split(os.pathsep):
+        exe_file = os.path.join(path, program)
+        if os.path.isfile(exe_file) and os.access(exe_file, os.X_OK):
+            return exe_file
+    return None
+
+def config_validator(config):
+    if not "name" in config:
+        logger.warning("name is not defined")
+    if not "scope" in config:
+        logger.warning("scope is not defined")
+    if not "postgresql" in config:
+        logger.warning("postgresql section is missing")
+    else:
+        if not "authentication" in config["postgresql"]:
+            logger.warning("postgresql.authentication is nor defined")
+        else:
+            if not "replication" in config["postgresql"]["authentication"]:
+                logger.warning("postgresql.authentication.replication is not defined")
+            if not "superuser" in config["postgresql"]["authentication"]:
+                logger.warning("postgresql.authentication.superuser is not defined")
+
+        if not "bin_dir" in config["postgresql"]:
+            logger.info("postgresql.bin_dir is not defined")
+        else:
+            if os.path.exists(config["postgresql"]["bin_dir"]):
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    exe_file = os.path.join(config["postgresql"]["bin_dir"], program)
+                    if not (os.path.isfile(exe_file) and os.access(exe_file, os.X_OK)):
+                        logger.warning("Program '%s' not found.", program)
+            else:
+                logger.warning("Directory '%s' does not exist.", config['postgresql']['bin_dir'])
+                for program in ["pg_ctl", "initdb", "pg_controldata", "pg_basebackup"]:
+                    if not which(program):
+                        logger.warning("Program '%s' not found.", program)
```

Comment:
There is quite a few duplicated code.
It could be simplified as:
```python
path = config['postgresql'].get('bin_dir') or None
for ...
  if not find_executable(program, path):
    logger.fatal...
```

Off course we also have to check that the `bin_dir` is a string if it is defined.